pkg/cloud/azure/services/compute/virtualmachines.go

```diff
@@ -42,14 +63,174 @@ func (s *Service) VMIfExists(resourceGroup string, name string) (*compute.Virtua
+				return input, err
+			}
+		} else {
+			klog.V(2).Infof("Machine %q is the first controlplane machine for cluster %q", machine.Name(), s.scope.Name())
+			if !s.scope.ClusterConfig.CAKeyPair.HasCertAndKey() {
+				return nil, errors.New("failed to run controlplane, missing CAPrivateKey")
+			}
+
+			cfg, err = config.NewControlPlane(&config.ControlPlaneInput{
+				CACert:           string(s.scope.ClusterConfig.CAKeyPair.Cert),
+				CAKey:            string(s.scope.ClusterConfig.CAKeyPair.Key),
+				EtcdCACert:       string(s.scope.ClusterConfig.EtcdCAKeyPair.Cert),
+				EtcdCAKey:        string(s.scope.ClusterConfig.EtcdCAKeyPair.Key),
+				FrontProxyCACert: string(s.scope.ClusterConfig.FrontProxyCAKeyPair.Cert),
+				FrontProxyCAKey:  string(s.scope.ClusterConfig.FrontProxyCAKeyPair.Key),
+				SaCert:           string(s.scope.ClusterConfig.SAKeyPair.Cert),
+				SaKey:            string(s.scope.ClusterConfig.SAKeyPair.Key),
+				// TODO: Fix LBAddress to retrieve LB DNS Name
+				LBAddress:         "fakelb.example.com", //s.scope.Network().APIServerIP.IPAddress,
+				ClusterName:       s.scope.Name(),
+				PodSubnet:         s.scope.Cluster.Spec.ClusterNetwork.Pods.CIDRBlocks[0],
+				ServiceSubnet:     s.scope.Cluster.Spec.ClusterNetwork.Services.CIDRBlocks[0],
+				ServiceDomain:     s.scope.Cluster.Spec.ClusterNetwork.ServiceDomain,
+				KubernetesVersion: machine.Machine.Spec.Versions.ControlPlane,
+			})
+
+			if err != nil {
+				return input, err
+			}
+		}
+
+		input.StartupScript = cfg
+
+	case "node":
+		// TODO: Check for existence of node subnet & ensure NSG is attached to subnet
+
+		cfg, err := config.NewNode(&config.NodeInput{
+			CACertHash:     caCertHash,
+			BootstrapToken: bootstrapToken,
+			// TODO: Fix LBAddress to retrieve LB DNS Name
+			LBAddress: s.scope.Network().APIServerIP.IPAddress,
+		})
+
+		if err != nil {
+			return input, err
+		}
+
+		input.StartupScript = cfg
+
+	default:
+		return nil, errors.Errorf("Unknown node role %q", machine.Role())
+	}
+
+	// TODO: Set ssh key
+
+	resourcesSvc := resources.NewService(s.scope)
+	err = resourcesSvc.ValidateDeployment(machine.Machine, s.scope.ClusterConfig, machine.MachineConfig, input.StartupScript)
+	if err != nil {
+		return nil, fmt.Errorf("error validating deployment: %v", err)
+	}
+
+	deploymentsFuture, err := resourcesSvc.CreateOrUpdateDeployment(machine.Machine, s.scope.ClusterConfig, machine.MachineConfig, input.StartupScript)
+	if err != nil {
+		return nil, fmt.Errorf("error creating or updating deployment: %v", err)
+	}
+
+	err = resourcesSvc.WaitForDeploymentsCreateOrUpdateFuture(*deploymentsFuture)
+	if err != nil {
+		return nil, fmt.Errorf("error waiting for deployment creation or update: %v", err)
+	}
+
+	deployment, err := resourcesSvc.GetDeploymentResult(*deploymentsFuture)
+	// Work around possible bugs or late-stage failures
+	if deployment.Name == nil || err != nil {
+		return nil, fmt.Errorf("error getting deployment result: %v", err)
```

Comment:
```suggestion
		return nil, errors.Wrapf(err, "error getting deployment result")
```
I love `Wrapf`. Let's use that everywhere for consistency since I see that you've already used it.

Reply:
Let's tackle this in a separate PR: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/94